### ISSUE TITLE
Enable enforcing password expiry

### ISSUE THREAD

**Issue by CrispinF** [cloudscribe/cloudscribe#1012]
Consider whether to enable a password expiry feature. 

Not recommended/needed when using MFA, but often required by organisations when not using MFA, despite general recommendations that password expiration is not a good thing (e.g. https://learn.microsoft.com/en-us/microsoft-365/admin/misc/password-policy-recommendations?view=o365-worldwide ).

Would need to be optional, activated e.g. by specifying a time period.

Logging in as the expiry data approaches (configurable days?) could show user a warning that password will expire at dd/mm/yyyy hh:mm:ss and show Change Password Now link. 

Attempting login after expiry should fail with the current feedback. The user can then use the 'Forgot your password?' link. Administrators can also reset passwords for users.

Should we check that a new password is not too similar to the previous one? Definitely check it's not the same. That's probably sufficient initially?

**Comment by markidox** [comment]
@JimKerslake PR for this is here: https://github.com/cloudscribe/cloudscribe/pull/1029

I haven't put these up to Nexus, but will need to test this with third-party logins as well as more generally.

**Comment by JimKerslake** [comment]
Yes the issue here would be what happens when you register with social auth (which I think creates a cs_user row with an empty password field) - 

At what point / mechanism can that user login be changed to being a CS one with a populated passwd field (?)  - and then at what point this interacts with the password changing requirement.  Clearly don't want to prompt a user to change their cs password when they haven't got one because they're using social auth only.

**Comment by JimKerslake** [comment]
OK verified that for a user who has registered with a social auth account only (and not set their own CS password - so that remains null in the Db) then the code for implementing this feature looks as though it will never enforce password changing rules. 

Need to ask similar questions for an AD authentication setup...

(( Also noticed that in the former scenario the user still gets a "change password" link in their nav menu - which is incorrect because they can't.  But that's a separate issue. ))

**Comment by JimKerslake** [comment]
```
  if (template.User.LastPasswordChangeUtc != null)
            {
```

I might be tempted to do belt & braces here and check that PasswordHash is not null as well as that other field before doing the nag.

**Comment by JimKerslake** [comment]
I guess we should also consider whether it ever makes any sense to have 
Password expiry alert display (days)
as a greater number than 
Number of days until password expires

With this combination I never seem to see the alert at all:
![image](https://github.com/user-attachments/assets/2bf6ff39-8c08-4521-a957-e7951c899d5d)

That's not a very sensible config, but does it need validation?
In reality I guess it's more like a 5 day warning window on a 180 day expiry time.

I'm getting a slight sense that these date time calculations are not always quite on the money here, since with a 4 day expiry, testing at 4pm on 28 Oct,  
when my last password change was  2024-10-23 03:44:30  then I fail to login (correct)  but  when my last password change was  2024-10-23 19:44:30   then I succeed (?).

**Comment by JimKerslake** [comment]
Here's a fairly major question:  should password expiry apply to *all* users, including administrators?  

There is already special workaround code in CS to prevent admins from locking themselves out of their own site...

Needless to say, in testing, I locked myself out (since admin was created a few days ago) and could not get back in again without hacking the Db.  

I suppose the "forgot my password" route could still remedy that?  (needs testing).

**Comment by JimKerslake** [comment]
Tested and confirmed that the password change screen disallows re-setting of password to same as the old one.

In passing... the "forgot password" route does allow that... (and interestingly the passwordhash field is different every time you re-set to the same passwd...) but that may be acceptable.

**Comment by JimKerslake** [comment]
Ah... there was a bug in the time calculations which meant we would never see this message... fixed that... but we also need to pull these text strings into ResX entries rather than inline in the view. 
![image](https://github.com/user-attachments/assets/c4738140-7408-449f-9324-99321dd70531)

**Comment by JimKerslake** [comment]
...although looking at it again, this  DefaultAccountLoginRulesProcessor  class is packed full of message strings that have never been abstracted into ResX...   Nothing in cloudscribe.Core.Identity ever has been.  Maybe there's a deeper reason for that...

**Comment by JimKerslake** [comment]
cloudscribe.Core.CompiledViews.Bootstrap5 8.0.15
cloudscribe.Core.Identity 8.0.7
cloudscribe.Core.Models 8.0.6
cloudscribe.Core.Web  8.0.16

cloudscribe.Core.Storage.EFCore.MSSQL 8.0.6
cloudscribe.Core.Storage.EFCore.PostgreSql 8.0.5
(and bump the other Dbs too)

Released but could do with some more pedantic testing really.
